llvm/include/llvm/Object/SYCLBIN.h

```diff
@@ -0,0 +1,130 @@
+//===- SYCLBIN.h - SYCLBIN binary format support ----------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_OBJECT_SYCLBIN_H
+#define LLVM_OBJECT_SYCLBIN_H
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/SYCLPostLink/ModuleSplitter.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include <string>
+
+namespace llvm {
+
+namespace object {
+
+// Representation of a SYCLBIN binary object. This is intended for use as an
+// image inside a OffloadBinary.
+class SYCLBIN {
+public:
+  SYCLBIN(MemoryBufferRef Source) : Data{Source} {}
+
+  SYCLBIN(const SYCLBIN &Other) = delete;
+  SYCLBIN(SYCLBIN &&Other) = default;
+
+  SYCLBIN &operator=(const SYCLBIN &Other) = delete;
+  SYCLBIN &operator=(SYCLBIN &&Other) = default;
+
+  MemoryBufferRef getMemoryBufferRef() const { return Data; }
+
+  enum class BundleState : uint8_t { Input = 0, Object = 1, Executable = 2 };
+
+  struct ModuleDesc {
+    BundleState State;
+    std::string ArchString;
+    std::vector<module_split::SplitModule> SplitModules;
+  };
+
+  /// The current version of the binary used for backwards compatibility.
+  static constexpr uint32_t CurrentVersion = 1;
+
+  /// Magic number used to identify SYCLBIN files.
+  static constexpr uint32_t MagicNumber = 0x53594249;
+
+  /// Serialize the contents of \p ModuleDescs to a binary buffer to be read
+  /// later.
+  static Expected<SmallString<0>> write(const SmallVector<ModuleDesc> &);
```

Comment:
I would discourage usage of `SmallString` here since `SmallVector`'s size is bounded by ~4GiB due to used `unsigned` type. This can be reached by big clients such as oneMKL or AMR. Especially, in debug mode.

You can calculate the output's size, then open a [WritableMemoryBuffer](https://llvm.org/doxygen/classllvm_1_1WritableMemoryBuffer.html) and then write your content there.

Reply:
Or simple `void print(raw_ostream &OS)`.